lib/reader/src/parser.rs

```diff
@@ -1086,8 +1104,29 @@ impl<'a> Parser<'a> {
         // Pass a reference to `ctx.values` instead of `ctx` itself since the `Values` iterator
         // holds a reference to `ctx.function`.
         self.add_values(&mut ctx.map,
-                        results.into_iter(),
-                        ctx.function.dfg.inst_results(inst))
+                        results.iter().map(|v| *v),
+                        ctx.function.dfg.inst_results(inst))?;
+
+        if let Some(ref result_locations) = result_locations {
+            if result_locations.len() != results.len() {
+                return err!(self.loc,
+                            "must have same number of result locations as results");
+            } else {
+                for (&src_value, &loc) in results.iter().zip(result_locations) {
+                    // We are safe to unwrap here because all values should have been added to the
+                    // map in the above call to self.add_values()
+                    let value = ctx.map.get_value(src_value).unwrap();
+                    *ctx.function.locations.ensure(value) = loc;
+                }
+            }
+        } else {
+            for &src_value in results.iter() {
+                let value = ctx.map.get_value(src_value).unwrap();
+                *ctx.function.locations.ensure(value) = ValueLoc::Unassigned;
```

Comment:
I pushed a fix for your crash in bc1901b766782176c9c2edede466de09387eee15. It is related to what you're doing here: The `write_function` code checks `if !func.locations.is_empty()`, and here we're filling default values into the map so it is no longer empty. It's a subtle difference, is the locations map empty or full of default values.

It affects if `write_function` will write a bunch of default value locations, or nothing at all.